components/utils.py

```diff
@@ -310,26 +336,72 @@ def yum_install(source):
         source_name = subprocess.check_output(
             ['rpm', '-qp', source_path]).strip()
 
-    ctx.logger.info('Checking whether {0} is already installed...'.format(
-        source_path.rstrip('\n\r')))
-    installed = run(['rpm', '-q', source_name], ignore_failures=True)
-    if installed.returncode == 0:
-        ctx.logger.info('Package {0} is already installed.'.format(source))
-        return
+        ctx.logger.info('Checking whether {0} is already installed...'.format(
+            source_path))
+        installed_rpm = _get_installed_rpm_source(source_name)
+        if installed_rpm == source_name:
+            ctx.logger.info('Package {0} is already installed.'.format(source))
+            return
+
+        if is_upgrade() or is_rollback():
+            _remove_existing_rpm_package(source_name)
```

Comment:
`yum update package.rpm` implements updating a package. Don't think we need to uninstall it first. you can also use `rpm -Uvh package.rpm` if needed.

instead if we call the relevant yum subcommand according to the context (bootstrap vs. upgrade)